### PR TITLE
use _getSettingsWithExactBounds in fewer cases

### PR DESCRIPTION
**Pull request by vmitnick**
Improve speed of getting a pixmap, by reducing the cases in which we need to do _getSettingsWithExactBounds

Squashed commit of the following:
commit d3843d7780093abec893ed51a0f6fa3340b5944f
Merge: 0a40204 90a3584
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Wed Oct 21 15:28:23 2015 -0700

    Merge branch 'master' into vmitnick/layer-effect-optimiz

commit 0a402045a78feddaf70d428f9ede72be13290233
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Wed Oct 21 10:32:24 2015 -0700

    Fix a half-typo that was left over after merging with master.

commit 3578c4879e756b219f3ac66565b348b80271a3ea
Merge: a39e635 ced18ed
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Tue Oct 20 16:49:26 2015 -0700

    Merge branch 'master' into vmitnick/layer-effect-optimiz

commit a39e6357f88adbe433ae3b1a25a2ac931bb84431
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Tue Oct 20 15:36:58 2015 -0700

    Remove a couple debugging log statements.

commit 6e6818aea2ea518091c056f7528a1d0913d24193
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Tue Oct 20 12:58:42 2015 -0700

    In DocumentManager, add new function clearDeferred(), which can be used
    when the caller (e.g. devicePreview) needs to be sure that there are no
    lingering deferred promises for getDocument.
    In PixmapRenderer.prototype._getData(), clean up hasComplexTransform a bit, to include the check
    for existence of layer.

commit 80945a3296a59e41d5ea9193e6c78e61623d78cc
Author: Vic Mitnick <vmitnick@adobe.com>
Date:   Mon Oct 19 14:53:27 2015 -0700

    In _componentHasLayerEffects(), ignore layerEffects for layers that
    are directly or indirectly invisible, that can't make the layer bigger, or
    that don't cause the layer to cross the bounds of its container (document
    or artboard).
    In _getData(), only be concerned with hasComplexTransform for layers/layer groups,
    not for whole documents.

### CODE REVIEW

**Comment by chrisbank** [comment]
reviewing, will squash commit too to get ready for merge

**Comment by chrisbank** [comment]
going to create a new branch since I can't easily squash around the merge

**Comment by chrisbank** [comment]
clean version of pull is here https://github.com/adobe-photoshop/generator-assets/pull/377

File: lib/renderer.js [hunk 1]
```diff
@@ -1164,10 +1183,12 @@
         // 4. The "include-ancestor-masks" config option is NOT set
         // 5. None of the layers has zero bounds. Sometimes they aren't computed and set to all 0's
         // 6. The layer is clipped, in which case layer.bounds is the clipped size and not the layer size
-        var hasComplexTransform = (component.hasOwnProperty("scale") && component.scale % 1 !== 0) ||
-                component.hasOwnProperty("width") || component.hasOwnProperty("height"),
+        var layer = component.layer,
+            hasComplexTransform = layer &&
```
Comment (mcilroyc) [inline lib/renderer.js]: Hey @vmitnick or @chrisbank - I believe this restriction (that `hasComplexTransform` can only be true for layer components) is preventing layer comps from using "exact bounds" ... which means that you can not scale layer comps with explicit bounds: Like "200pxx200px Layer Comp.png"

Any memories about this? 

I see the commit msg 
>In _getData(), only be concerned with hasComplexTransform for layers/layer groups, not for whole documents.

Perhaps we should loosen that to "layers, layer groups, or layer comps"?

Is it a reasonable assumption that we don't want to allow explicit height/width for "whole documents"?
cc: @iwehrman

Comment (chrisbank) [inline lib/renderer.js]: I'm not clear when you can't scale in this case. The main reason for this claus is for the case where it's possible the cached bounds information for getDocumentInfo is not correct. The comps should have the same bounds as document.bounds which should always be correct. This is just, should be pre-generate the image at 100% to know it's bounds for sure or can safely calculate it off the information we already have.

Comment (mcilroyc) [inline lib/renderer.js]: I'll be honest @chrisbank, I don't really understand your response.  But I created a PR with my very limited fix, so if you have a moment to check that out, please do.  https://github.com/adobe-photoshop/generator-assets/pull/393

Comment (chrisbank) [inline lib/renderer.js]: It's basically, I think the bug is in _getSettingsWithApproximateBounds instead. This is basically forcing all layer comps into the slow bounds calculation case. In the code there's this clause:

https://github.com/adobe-photoshop/generator-assets/blob/ea884529c40f5b3752273fad4a88be17f0a9b207/lib/renderer.js#L862

The bounds for the layer comp should be the same as the bounds for the regular document. What is the output you're seeing from both cases?

Comment (mcilroyc) [inline lib/renderer.js]: I see.  You may be right that it is a limitation with `_getSettingsWithApproximateBounds `, but specifically it is that in that method there is no consideration of an explicit height/width, only providing for scalar values.  see [here](https://github.com/adobe-photoshop/generator-assets/blob/ea884529c40f5b3752273fad4a88be17f0a9b207/lib/renderer.js#L892-L894) vs the bit in exact bounds [here](https://github.com/adobe-photoshop/generator-assets/blob/ea884529c40f5b3752273fad4a88be17f0a9b207/lib/renderer.js#L800-L801)

So, I think your point is that I am missing an opportunity to optimize for layer comps by updating `_getSettingsWithApproximateBounds` to handle height/width calculations based on the less expensive data.  My plea for my newest PR (393) is that we are just reverting back to the functionality before this PR(374), and consider this optimization in the future.

To answer your question, the badness I'm seeing now is that layer comps which are scaled (via the image assets plugin) with an explicit height/width (such as `200x200 comp1.png`) are getting output *at original dimension*.

**Pull request closed without merging**